### DISCUSSION TITLE
How to train and deploy an ultra-small YOLO model on a constrained device (OpenArtPlus v1.0, 2MB SRAM, 64MB SDRAM)?

### DISCUSSION THREAD

**Discussion by GaoPeng415** [ultralytics/ultralytics#23315] [Q&A]
Hello everyone,  

I am trying to train a model for my smart car to recognize **tennis balls** and **teddy bears**—this is a requirement for the 21st Smart Car Competition. More details about the competition can be found here: [link](https://zhuoqing.blog.csdn.net/article/details/154691441?spm=1011.2415.3001.5331).  

We have a camera module (OpenArtPlus v1.0) which has **2MB SRAM** and **64MB SDRAM**, and supports **TensorFlow Lite**. You can find its specifications here: [NXP AIoT Cloud](https://aiotcloud.nxp.com.cn/product-smart-car/open-art-series) and [Gitee](https://gitee.com/seekfree/OpenART_plus_Product).  

I have tried using the following YOLO models: `yolo26n.pt`, `yolov5nu.pt`, `yolov8n.pt`, `yolov3tinyu.pt` to train my model and export to TFLite with `int8=True`. However, the smallest final model is **2624KB**, which cannot run properly on the camera. Meanwhile, the example model provided by the OpenArtPlus repository is only **103KB**, which surprised me—I cannot figure out how it is so small.  

I am new to YOLO and only have a basic understanding. Could anyone kindly give me guidance or advice on how to train a model for recognizing **tennis balls** and **teddy bears** that can run on this constrained device? My ultimate goal is to use the model to locate objects and control the smart car to lift them.  

Thank you very much for your help!  

P.S. I’m not sure if asking this kind of help here aligns with the community guidelines. If this post is inappropriate, please let me know and I will remove it. If possible, I’d appreciate suggestions on where it would be more appropriate to ask for help.  =)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @GaoPeng415, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get moving quickly, and an Ultralytics engineer will also assist soon 😊

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (your constrained-device + TFLite export scenario sounds like it), please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
With only 2MB SRAM, the blocker is usually **runtime RAM (activations + TFLite buffers + postprocess/NMS)** rather than the `.tflite` weight file size, so an INT8 export in the ~2–3MB range is expected for standard YOLO detectors and still may not run; that ~103KB “example model” is very likely a tiny classifier or a vendor-specific packed artifact (not a YOLO-style detector), so it’s not an apples-to-apples comparison. If you still want to push YOLO as far as possible, try exporting with a much smaller `imgsz` (this won’t shrink weights much, but it can drastically reduce peak RAM) and make sure you’re doing a calibrated INT8 export as shown in the [TFLite export guide](https://docs.ultralytics.com/integrations/tflite/) (also see [model quantization notes](https://www.ultralytics.com/glossary/model-quantization)):

```python
from ultralytics import YOLO

model = YOLO("yolo26n.pt")
model.export(format="tflite", int8=True, data="your_data.yaml", imgsz=320)
```

If you can paste the exact on-device failure log (text) and share what that 103KB model file actually is (format + how it’s loaded), we can better tell whether you’re hitting SRAM limits, unsupported ops, or a postprocessing constraint.

**Reply by GaoPeng415** [reply]
Thank you very much for your advice, and I apologize for the late reply. I was busy preparing for the competition and unfortunately missed this notification.

Over the past few months, I have learned much more about vision models, so I can now provide more details about the 103 KB model. I would really appreciate any further advice you may have.

You were absolutely right that the 103 KB model is not a traditional YOLO model. Its architecture is shown here:

[MODEL_ARCHITECTURE.md](https://github.com/user-attachments/files/27298446/MODEL_ARCHITECTURE.md)

In short, it is not a standard YOLO model, but rather a very lightweight detector built with a MobileNetV2-style backbone and a YOLO-style head. That is why it is much smaller than standard YOLO models.

With some modifications of my own, I can now use this model structure to roughly complete my visual recognition task, and the detected bounding boxes already meet my needs. Specifically, I trained it with about 10,000 images in total, covering four classes with roughly 2,500 samples per class. The model was trained at a resolution of 96x96, and the images to be detected are first letterboxed before being fed into the model.

If you have any further suggestions on how to achieve the best possible performance, I would be very grateful. In particular, I would appreciate advice on both model architecture improvements and training strategy improvements.

**Reply by glenn-jocher** [reply]
Thanks for the extra detail — for a `96x96` detector on `2MB` SRAM, the best gains usually come from simplifying the model more than training longer: keep a single detection scale, use depthwise-separable blocks, remove heavy neck/FPN stages, minimize channels/anchors/classes, and prefer INT8 with quantization-aware training plus light channel pruning, which matches the ideas in our [deployment optimization guide](https://docs.ultralytics.com/guides/model-deployment-practices) and [hardware-aware speed overview](https://www.ultralytics.com/blog/how-to-make-yolo-models-fast-on-your-favorite-chip). On training, train exactly at deployment resolution, bias sampling toward your hardest small-object cases, reduce strong augmentation near the end, and evaluate by on-device latency/RAM as much as mAP; if you share a few failure cases and your current FPS/RAM numbers, we can suggest more targeted tweaks.
